aws-android-sdk-cognitoidentityprovider/src/main/java/com/amazonaws/mobileconnectors/cognitoidentityprovider/CognitoUser.java

```diff
@@ -720,11 +730,14 @@ protected CognitoUserSession getCachedSession() {
             try {
                 cipSession = refreshSession(cachedTokens);
                 cacheTokens(cipSession);
+                CognitoUser.semaphore.release();
                 return cipSession;
             } catch (NotAuthorizedException nae) {
                 clearCachedTokens();
+                CognitoUser.semaphore.release();
                 throw new CognitoNotAuthorizedException("User is not authenticated", nae);
             } catch (Exception e) {
+                CognitoUser.semaphore.release();
```

Comment:
Since we are dealing with locking here, can you move the release into a finally block?